### PR TITLE
chore: created new type for API errors, created and implemented type predicates

### PR DESCRIPTION
**Pull request by dylanspyer**
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes conversion to TypeScript

Previously, most invocations of the custom `error` function needed to have a TypeScript ignore comment due to the `error_` object being passed to the function being of type `unknown`. Because the `error` function already takes `error_` arguments that are of various different shapes, adding `unknown` to the parameter's type union allowed us to remove these TypeScript ignore comments without sacrificing type safety. Added a new type `APIError` and type predicates for custom type guarding of errors that are a result of an internal API call.

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/cli/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

### CODE REVIEW

File: src/utils/command-helpers.ts [hunk 1]
```diff
@@ -300,3 +299,14 @@ export const nonNullable = <T>(value: T): value is NonNullable<T> => value !== n
...
+interface APIError extends Error {
+  status: number
+  message: string
+}
+
+export const isAPIError = (errorObject: unknown): errorObject is APIError =>
+  errorObject instanceof Error && 'status' in errorObject && 'message' in errorObject
+
+export const errorHasStatus = (errorObject: unknown, statusCode: number) =>
+  isAPIError(errorObject) && errorObject.status === statusCode
```
Comment (DanielSLew) [inline src/utils/command-helpers.ts:312]:
```suggestion
export interface APIError extends Error {
  status: number
  message: string
}
```

File: src/utils/dev.ts [hunk 2]
```diff
@@ -63,8 +64,9 @@ const getAddons = async ({ api, site }) => {
     const addons = await api.listServiceInstancesForSite({ siteId: site.id })
     return addons
   } catch (error_) {
-    // @ts-expect-error TS(2571) FIXME: Object is of type 'unknown'.
-    error(`Failed retrieving addons for site ${chalk.yellow(site.id)}: ${error_.message}. ${ERROR_CALL_TO_ACTION}`)
+    isAPIError(error_)
+      ? error(`Failed retrieving addons for site ${chalk.yellow(site.id)}: ${error_.message}. ${ERROR_CALL_TO_ACTION}`)
+      : error(error_)
```
Comment (DanielSLew) [inline src/utils/dev.ts:69]:
```suggestion
        error(`Failed retrieving addons for site ${chalk.yellow(site.id)}: ${(error_ as APIError).message}. ${ERROR_CALL_TO_ACTION}`)

```

File: src/utils/framework-server.ts [hunk 3]
```diff
@@ -70,7 +70,7 @@ export const startFrameworkServer = async function ({
     stopSpinner({ error: true, spinner })
     log(NETLIFYDEVERR, `Netlify Dev could not start or connect to localhost:${settings.frameworkPort}.`)
     log(NETLIFYDEVERR, `Please make sure your framework server is running on port ${settings.frameworkPort}`)
-    // @ts-expect-error TS(2345) FIXME: Argument of type 'unknown' is not assignable to pa... Remove this comment to see the full error message
+    //
```
Comment (DanielSLew) [inline src/utils/framework-server.ts:73]:
```suggestion
```

File: src/utils/hooks/requires-site-info.ts [hunk 4]
```diff
@@ -15,17 +16,14 @@ const requiresSiteInfo = async (command) => {
     await api.getSite({ siteId })
   } catch (error_) {
     // unauthorized
-    // @ts-expect-error TS(2571) FIXME: Object is of type 'unknown'.
-    if (error_.status === 401) {
+    if (errorHasStatus(error_, 404)) {
```
Comment (DanielSLew) [inline src/utils/hooks/requires-site-info.ts:19]:
```suggestion
    if ((error_ as APIError).status === 401) {
```

**Review by DanielSLew — CHANGES_REQUESTED** [review]
I think this is a simpler version of adding the type and it's less for the users to remember to implement as they add more error catching.

**Comment by dylanspyer** [comment]
@DanielSLew Changed to type assertions for APIError

**Pull request merged**